src/components/chat/UserList.js

```diff
@@ -11,80 +11,69 @@ class UserList extends Component {
   }
 
   styleUserList = () => {
-    if(this.props.displayUserList) {
-      return { display: 'block'};
+    if (this.props.displayUserList) {
+      return { display: 'block' };
     } else {
-      return { display: 'none'};
+      return { display: 'none' };
     }
   }
 
-  sortByFrom(status1, status2) {
-    return status1.from.toLowerCase().localeCompare(status2.from.toLowerCase());
+  sortByFrom(username1, username2) {
+    return username1.toLowerCase().localeCompare(username2.toLowerCase());
   }
 
   render() {
-    let statuses = this.props.statuses;
-
-    let viewing = [];  // green
-    let online = [];   // yellow
-    let offline = [];  // gray
-
-    let usernamesSeen = [];
-
-    for (let i = statuses.length - 1; i >= 0; i--) {
-      let status = statuses[i];
-
-      // Avoid showing a user's status twice
-      if (usernamesSeen.indexOf(status.from) !== -1) {
-        continue;
+    const { statuses } = this.props;
+
+    const viewing = [];  // green
+    const online = [];   // yellow
+    const offline = [];  // gray
+
+    Object.keys(statuses).forEach(username => {
+      const status = statuses[username];
+      switch (status) {
+        case 'viewing':
+          viewing.push(username);
+          break;
+        case 'online':
+          online.push(username);
+          break;
+        case 'offline':
+          offline.push(username);
+          break;
       }
-      switch (status.status) {
-      case 'viewing':
-        viewing.push(status);
-        break;
-      case 'online':
-        online.push(status);
-        break;
-      case 'offline':
-        offline.push(status);
-        break;
-      }
-      console.log('Already seen', status.from);
-      usernamesSeen.push(status.from);
-    }
+    });
 
     viewing.sort(this.sortByFrom);
     online.sort(this.sortByFrom);
     offline.sort(this.sortByFrom);
 
-    console.log('Updating presence/all user statuses:', viewing, online, offline);
-
     return (
       <div className="users-list">
         <ul style={this.styleUserList()}>
-          {viewing.map(status => {
+          {viewing.map((username, i) => {
             return (
-              <li key={status.key}>
+              <li key={i}>
```

Comment:
@Spetastian Do keys not have to be globally unique? just unique to the component they're used within?

Reply:
From the docs:

> The best way to pick a key is to use a string that uniquely identifies a list item among its siblings. Most often you would use IDs from your data as keys: ...

> When you don't have stable IDs for rendered items, you may use the item index as a key as a last resort: ...

Awesome, didn't know that :-).

Reply:
Best thing would be if we hade consistent unique IDs for messages to use. To generate a unique identifier as we did before on the client every time message is received is basically the same as using the index as a key.

Reply:
Each message has a UUID that we will soon expose to the frontend :-)